---
title: "Nanopolish FVB Resolution"
author: "Scott Gigante"
date: "20/07/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(viridis)
library(forcats)
library(gridExtra)

```

```{r}
min_coverage=5
base_fn <- "../nanopore/b6xcast.minion"
haplotype_df <- read_tsv(paste0(base_fn, ".threeway_phased.tsv")) %>%
  rename(read_name=read) %>%
  inner_join(read_tsv(paste0(base_fn, ".summary.tsv"), 
                      col_names=c("read_name", "chr", "start", "end", "qual"), skip=1))
```

## Define functions

```{r fig.height=10, fig.width=10}
assign_chr_genotype <- function(.chr, df, binwidth=1e5, smooth=10) {
  df <- df %>% dplyr::filter(chr==.chr, genotype %in% c("ref", "alt2"))
  result <- data_frame(pos=seq(from=binwidth/2, to=max(df$end)-binwidth/2, by=binwidth/2))
  result$prop <- sapply(result$pos, function(i) {
    .start = i-binwidth/2
    .end=i+binwidth/2
    df <- df %>% dplyr::filter(start >= .start, end < .end)
    sum(df$genotype == "ref")/nrow(df)
  })
  tree <- rpart::rpart(prop ~ pos, data=result, control=rpart::rpart.control(minsplit=5, cp=.1))
  result$tree <- predict(tree, data.frame(pos=result$pos))
  result$chr <- as.character(.chr)
  result
}

plot_rpart <- function(partitions) {
  partitions %>%
    bind_rows() %>%
    filter(chr!="Y") %>%
    mutate(pos = pos/1e6,
           prop=1-prop,
           tree=1-tree,
           chr=paste0("chr", chr),
           chr=fct_relevel(as.factor(chr), 
                           "chr1", "chr2", "chr3", "chr4", 
                           "chr5", "chr6", "chr7", "chr8", "chr9")) %>%
    ggplot(aes(x=pos)) +
    geom_point(aes(y=prop), alpha=0.05, na.rm=TRUE) +
    geom_line(aes(y=tree), color='red') +
    facet_wrap(~ chr, ncol = 4, scales="free_x") +
    labs(x="Genomic Position (Mb)",
         y="FVB Proportion of Maternal Reads")
}

rpart_to_region <- function(result) {
  split_idx <- which(result$tree[1:(nrow(result)-1)] - result$tree[2:nrow(result)] != 0)
  if (length(split_idx)==0) {
    return()
  }
  regions <- c()
  last_pos <- 0
  for (i in 1:length(split_idx)) {
    curr_pos <- (result$pos[split_idx[i]]+result$pos[split_idx[i]+1])/2
    if (result$tree[split_idx[i]] < 0.6) {
      regions <- c(regions, paste(last_pos, curr_pos, sep="-"))
    }
    last_pos <- curr_pos
  }
  if (result$tree[split_idx[length(split_idx)]+1] < 0.6) {
        regions <- c(regions, curr_pos)
  }
  if (length(regions) > 0) {
    paste(paste0("chr", result$chr[1]), regions, sep=":")
  }
}
```

## Figure S5

```{r fig.height=12, fig.width=10}
partitions <- lapply(c(as.character(1:19), "X", "Y"), assign_chr_genotype, haplotype_df)

plot_rpart(partitions)
ggsave("../plots/rpart_fvb_resolution.png")
```

## Output results as region string

```{r}
# chrY can't have anything, it's Cast only
partitions[[21]] <- NULL
fvb_regions <- lapply(partitions, rpart_to_region) %>%
  unlist() %>%
  paste(collapse=" ")
fvb_regions %>%
  write_file("../genome_data/fvb_regions.txt")
fvb_regions
```

