---
title: "RNAseq Differential Expression Analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(cowplot)
if (!require(ggrastr)) devtools::install_github("VPetukhov/ggrastr")
library(ggrastr)
library(tidyverse)
library(Rsubread)
library(limma)
library(edgeR)
library(biomaRt)
```

## Process bam files

```{r} 
#list bam files
splitbam <- list.files('../rna_seq/', pattern = '(CastB6|B6Cast)F1_[2-5].hisat2.genome[1-2].bam$')

#get counts
GTF <- '../genome_data/Mus_musculus.GRCm38_90.chr.gtf'

splitfeaturecounts <- featureCounts(file.path("..", "rna_seq", splitbam), annot.ext= GTF,
                                    isGTFAnnotationFile = TRUE, useMetaFeatures = TRUE, 
                                    primary = TRUE, isPairedEnd = TRUE, nthreads = 8)
splitcounts <- as.data.frame(splitfeaturecounts$counts)
colnames(splitcounts) <- splitbam %>%
    str_replace("B6CastF1_([2-5]).hisat2.genome1.bam", "B6CastF1_\\1_mat") %>%
    str_replace("B6CastF1_([2-5]).hisat2.genome2.bam", "B6CastF1_\\1_pat") %>%
    str_replace("CastB6F1_([2-5]).hisat2.genome1.bam", "CastB6F1_\\1_pat") %>%
    str_replace("CastB6F1_([2-5]).hisat2.genome2.bam", "CastB6F1_\\1_mat")
splitcounts$GeneID <- rownames(splitcounts)
write_tsv(splitcounts, '../rna_seq/hisat2_GCRm38V90_featurecounts_split.tsv.gz')
```


# Differential expression analysis

## imprinted (parent-biased) expression


```{r}
counts <- dplyr::select(splitcounts, -GeneID)
counts <- as.data.frame(counts)
rownames(counts) <- splitcounts$GeneID
#filter genes on X and Y chromosomes
# remove X and Y and MT and duplicated genes
mouse <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")
gene <- getBM(filters = 'ensembl_gene_id', attributes = c('mgi_symbol','ensembl_gene_id','chromosome_name', 'start_position', 'end_position'), mart = mouse, values = rownames(counts))
#differing number of rows: 3 more in gene.
#remove duplicated genes
gene <- gene[!duplicated(gene$ensembl_gene_id),]
#select genes in chr 1 to 19 only
sel <- subset(gene, chromosome_name %in% as.character(1:19))
counts <- counts[rownames(counts) %in% sel$ensembl_gene_id,]	#48,242k genes
#filter genes with 0 counts:
counts <- counts[rowSums(counts)>0,]
#29859 genes left
```

```{r}
splitDE <- DGEList(counts = counts)
#TMM norm
splitDE <- calcNormFactors(splitDE)

embryo <- colnames(counts) %>%
  str_replace("_(mat|pat)", "") %>%
  factor()
strain <- ifelse((startsWith(colnames(counts), "B6") & endsWith(colnames(counts), "mat")) | 
                   (startsWith(colnames(counts), "Cast") & endsWith(colnames(counts), "pat")), 
                 "Black6", "Cast") %>%
  factor()
litter <- case_when(str_detect(colnames(counts), "^B6CastF1_(2|3)") ~ "A",
                    str_detect(colnames(counts), "^B6CastF1_(4|5)") ~ "B",
                    str_detect(colnames(counts), "^CastB6F1_(2|3)") ~ "C",
                    str_detect(colnames(counts), "^CastB6F1_(4|5)") ~ "D") %>%
  factor()
parent <- ifelse(endsWith(colnames(counts), "mat"), "maternal", "paternal") %>%
  factor(levels = c('paternal','maternal')) #change that so that higher expr will mean maternal and will be red.

colData <- data.frame(parent = parent, strain = strain, litter = litter, embryo = embryo)

design <- model.matrix(~embryo+parent)  #account for paired data

splitDE <- estimateDisp(splitDE, design)

splitfit <- glmQLFit(splitDE, design)

qlf <- glmQLFTest(splitfit)

FDR <- p.adjust(qlf$table$PValue, method="BH")

dt.edger <- decideTestsDGE(qlf, p.value =0.1) #default adjust method is BH, and will pick p.value = 0.05

#make annotation dataframe
annot <- counts %>% 
  rownames_to_column('ensembl_gene_id') %>% 
  as_tibble() %>% 
  left_join(gene) %>% 
  rename(ID = ensembl_gene_id) %>% 
  as.data.frame()

imprint_de_results <- qlf$table %>% 
  rownames_to_column('ensembl_gene_id') %>% 
  as_tibble() %>% 
  mutate(padj = FDR) %>% 
  left_join(gene) %>% 
  arrange(padj)

impr10 <- imprint_de_results %>% 
  dplyr::filter(padj < 0.1)
#135 genes.
#88 MEGs, 47 PEGs.
write_tsv(impr10, '../rna_seq/parent_biased_genes_10pctFDR.tsv')
write_tsv(imprint_de_results, '../rna_seq/imprint_DE_results.tsv.gz')
```

## Additional File 1

```{r}
#load data
load('../genome_data/known_imprinted_genes.RData')
#
annotated_imprinted <- impr10 %>% 
  mutate(in_db = ensembl_gene_id %in% 
           databases$ensembl_gene_id) %>% 
  mutate(in_andergassen = ensembl_gene_id %in% andergassen$ensembl_gene_id) %>% 
  left_join(dplyr::select(andergassen, ensembl_gene_id, known_ig, final_conclusion_on_imprinted_status)) %>% 
  mutate(in_Finn = ensembl_gene_id %in% finn$ensembl_gene_id) %>% 
  mutate(in_Calabrese = ensembl_gene_id %in% calabrese$ensembl_gene_id) %>% 
  mutate(in_Inoue = ensembl_gene_id %in% inoue$ensembl_gene_id) %>% 
  dplyr::select(-known_ig, -final_conclusion_on_imprinted_status)

write_tsv(annotated_imprinted, '../tables/imprinted_genes_Supp_file_1.tsv')
```

## strain-biased expression

```{r} 
cbbPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

strainDE <- DGEList(counts = counts)
#TMM norm
strainDE <- calcNormFactors(strainDE)

colData <- data.frame(parent = parent, strain = strain, litter = litter, embryo = embryo)

design2 <- model.matrix(~embryo+strain)  #account for paired data

strainDE <- estimateDisp(strainDE, design2)

strainfit <- glmQLFit(strainDE, design2)

qlfstrain <- glmQLFTest(strainfit)

FDRstrain <- p.adjust(qlfstrain$table$PValue, method="BH")


dt.edger.strain <- decideTestsDGE(qlfstrain, p.value = 0.05, lfc =1) #5% FDR and log2FC of 1
#make annotation dataframe
annot <- counts %>% rownames_to_column('ensembl_gene_id') %>% as_tibble() %>% left_join(gene) %>% rename(ID = ensembl_gene_id) %>% as.data.frame()

strain_de_results <- qlfstrain$table %>% rownames_to_column('ensembl_gene_id') %>% as_tibble() %>% mutate(padj = FDRstrain) %>% left_join(gene) %>% arrange(padj)

strain05log2 <- strain_de_results %>% dplyr::filter(padj < 0.05 & abs(logFC) > 1)

write_tsv(strain05log2, '../rna_seq/strain_biased_genes_5pctFDR.tsv')
write_tsv(strain_de_results, '../rna_seq/strain_DE_results.tsv.gz')
```

## Figure 3A

```{r} 
newgenes <- dplyr::filter(annotated_imprinted, in_db + in_andergassen + in_Finn + in_Calabrese + in_Inoue ==0)
```

```{r}
p <- imprint_de_results  %>% 
  mutate(status = case_when(padj < 0.1 & logFC > 0 ~ "Maternal dominant", 
                            padj < 0.1 & logFC < 0 ~ 'Paternal dominant', 
                            TRUE ~ 'Not significant')) %>% 
  mutate(status = factor(status, levels = c('Maternal dominant', 
                                            'Paternal dominant', 
                                            'Not significant')),
         new=ifelse(ensembl_gene_id %in% newgenes$ensembl_gene_id,
                    "Novel",
                    "Known"),
         size=ifelse(status=="Not significant", 1, 2))  %>%
  arrange(desc(status)) %>%
  ggplot(aes(x = logCPM, y = logFC, colour = status, shape=new, size=size)) + 
  geom_point_rast() + 
  labs(x = 'Average log CPM', 
       y = 'log2 Fold Change', 
       colour = 'Differential Expression', 
       title = 'Imprinted Expression',
       shape = 'Novel imprinted gene') + 
  theme_gray() + 
  scale_color_manual(values=c('Maternal dominant' = "#F8766D", 
                              'Paternal dominant' = "#00BFC4", 
                              'Not significant' = 'darkgrey')) + 
  scale_size_continuous(range=c(0.5, 3), guide='none') + 
  guides(color = guide_legend(order = 1), 
         shape = guide_legend(order = 2))
p
```

## Figure 3B

```{r}
p2 <- strain_de_results  %>% 
  mutate(status = case_when(padj < 0.05 & logFC > 1 ~ "Cast dominant", 
                            padj < 0.05 & logFC < -1 ~ 'B6 dominant', 
                            TRUE ~ 'Not significant')) %>% 
  mutate(status = factor(status, levels = c('Cast dominant', 
                                            'B6 dominant', 
                                            'Not significant')),
         new=ifelse(ensembl_gene_id %in% newgenes$ensembl_gene_id,
                    "Novel",
                    "Known"))  %>%
  arrange(desc(status)) %>%
  ggplot(aes(x = logCPM, y = logFC, colour = status)) + 
  geom_point_rast(size=0.5) + 
  labs(x = 'Average log CPM', 
       y = 'log2 Fold Change', 
       colour = 'Differential Expression', 
       title = 'Strain-biased Expression') + 
  theme_gray() + 
  scale_color_manual(values=c('Cast dominant' = "#E69F00", 
                              'B6 dominant' = "#000000", 
                              'Not significant' = 'darkgrey'))
p2
```

## Figure 3 combined

```{r}
#arrange both plots side by side:
p3 <- plot_grid(p, p2, labels = c("A", "B"), label_size=24)
p3
save_plot('../plots/MD-plots_combined.pdf', p3, base_aspect_ratio = 3)
```

# Figure S4

```{r} 
bias <- tibble(libsizes = splitDE$samples$lib.size, 
               libsizes_norm = splitDE$samples$lib.size * splitDE$samples$norm.factors, 
               cross = c(rep('B6xCast',8), rep('CastxB6',8)), 
               allele = c(rep(c('Maternal','Paternal'),4),rep(c('Paternal','Maternal'),4)), 
               embryo = c('162.5','162.5','162.8','162.8','163.4','163.4','163.6','163.6','618.1','618.1','618.4','618.4','627.5','627.5','627.6','627.6'))

p <- ggplot(bias, aes(x = embryo, y = libsizes/1e6, fill = allele)) + 
  geom_bar(stat='identity', position = 'dodge') + 
  facet_wrap(~ cross,scales = "free_x" ) + 
  ylim(c(0,NA)) + 
  labs(x = 'Sample', y = 'Counts (millions)', fill = 'Allele') + 
  theme_gray()
p
save_plot('../plots/allelic_bias.pdf', p, base_aspect_ratio = 2)
```
